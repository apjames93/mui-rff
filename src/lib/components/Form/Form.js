/* eslint-disable max-len */
import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import { Form as ReactFinalForm } from 'react-final-form';

export const RffWrapper = ({
  children,
  debug,
  initialValues,
  initialValuesEqual,
  keepDirtyOnReinitialize,
  onSubmit,
  subscription,
  validate,
  validateOnBlur,
  mutators,
}) => (
  <ReactFinalForm
    debug={debug}
    initialValues={initialValues}
    initialValuesEqual={initialValuesEqual}
    keepDirtyOnReinitialize={keepDirtyOnReinitialize}
    onSubmit={onSubmit}
    subscription={subscription}
    validate={validate}
    validateOnBlur={validateOnBlur}
    mutators={mutators}
  >
    {({ handleSubmit, submitting, form }) => (
      <form onSubmit={handleSubmit}>
        <>
          {Array.isArray(children) && children.map((child, key) => cloneElement(
            child,
            {
              key,
              submitting,
              form,
            },
          ))}

          {!Array.isArray(children) && cloneElement(
            children,
            {
              submitting,
              form,
            },
          )}
        </>
      </form>
    )}
  </ReactFinalForm>
);
export default RffWrapper;

// https://final-form.org/docs/react-final-form/types/FormProps
RffWrapper.propTypes = {
  /**
   * A render function that is given FormRenderProps, as well as any non-API props passed into the <Form/> component.
   */
  children: PropTypes.node.isRequired,
  /**
   * Function to call when the form is submitted. There are three possible ways to write an onSubmit function:
   */
  onSubmit: PropTypes.func.isRequired,

  /**
   * A callback for debugging that receives the form state and the states of all the fields. It's called on every state change. A typical thing to pass in might be console.log.
   */
  debug: PropTypes.func,
  /**
   * The initial values of your form. These will also be used to compare against the current values to calculate pristine and dirty.
   */
  initialValues: PropTypes.shape(),
  /**
   * A predicate to determine whether or not the initialValues prop has changed, i.e. to know if the form needs to be reinitialized with the new values. Useful for passing in a "deep equals" function if you need to. Defaults to "shallow equals".
   */
  initialValuesEqual: PropTypes.bool,
  /**
   * If true, only pristine values will be overwritten when initialize(newValues) is called. This can be useful for allowing a user to continue to edit a record while the record is being saved asynchronously, and the form is reinitialized to the saved values when the save is successful. Defaults to false.
   */
  keepDirtyOnReinitialize: PropTypes.bool,
  /**
   * An object of the parts of FormState to subscribe to. If a subscription is provided, the <Form/> will only rerender when those parts of form state change.
    If no subscription is provided, it will default to subscribing to all form state changes. i.e. <Form/> will rerender whenever any part of the form state changes.
   * ex : { [string]: boolean }
    */
  subscription: PropTypes.shape(),
  /**
   * ex: (values: FormValues) => Object | Promise<Object>
   * A whole-record validation function that takes all the values of the form and returns any validation errors. There are two possible ways to write a validate function:
   */
  validate: PropTypes.func,
  /**
   * If true, validation will happen on blur. If false, validation will happen on change. Defaults to false.
   */
  validateOnBlur: PropTypes.bool,
  /**
   * Optional.
   * Named mutator functions.
   */
  mutators: PropTypes.shape(),
};

RffWrapper.defaultProps = {
  debug: undefined,
  initialValues: {},
  initialValuesEqual: false,
  keepDirtyOnReinitialize: false,
  subscription: undefined,
  validate: undefined,
  validateOnBlur: false,
  mutators: undefined,
};
