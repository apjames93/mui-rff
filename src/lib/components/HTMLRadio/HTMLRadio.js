import React from 'react';
import PropTypes from 'prop-types';
import { FormControlLabel } from '@material-ui/core';
import RFFField from '../RFFField/RFFField';

export const HTMLRadio = ({
  name,
  value,
  disabled,
  label,
  labelPlacement,
  passProps,
}) => (
  <RFFField
    type="radio"
    name={name}
    value={value}
    passProps={passProps}
  >
    <FormControlLabel
      control={(
        <input
          data-testid="HTMLRadio"
          type="radio"
          disabled={disabled}
        />
      )}
      label={label}
      value={value}
      labelPlacement={labelPlacement}
    />
  </RFFField>
);

export default HTMLRadio;

HTMLRadio.propTypes = {
  /**
   * props to pass to react final form field in call back have access to field as first arg
   */
  passProps: PropTypes.func,
  /**
   * The name of your field. Field values may be deeply nested using dot-and-bracket syntax.
   */
  name: PropTypes.string.isRequired,
  /**
   * The label content.
   */
  label: PropTypes.node,
  /**
   * If true, the input element will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * A DOMString representing the value of the radio button.
   */
  value: PropTypes.string,
  /**
   * MUI Props: 'bottom' | 'end' | 'start' | 'top'
   * The position of the label.
   */
  labelPlacement: PropTypes.string,
};

HTMLRadio.defaultProps = {
  passProps: field => ({
    id: field.input.name,
    value: field.input.value,
    checked: field.input.checked,
    onChange: field.input.onChange,
  }),
  label: '',
  disabled: false,
  value: null,
  labelPlacement: 'end',
};
