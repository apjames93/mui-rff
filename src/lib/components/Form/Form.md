Form example:

```js noeditor
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';
```

```js
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';


const onSubmit = async (values) => {
  console.log('Input Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <MuiInput
      name="thing"
      label="thing"
    />
    <MuiInput
      name="number"
      label="number"
      type="number"
    />
    <MuiInput
      name="date"
      label="date"
      type="date"
    />
    <MuiSelect
      name="stuff"
      options={[{ num: 'six' }, { num: 'nine' }]}
      displayKey="num"
    />
    <MuiSubmit />
  </Form>
```
