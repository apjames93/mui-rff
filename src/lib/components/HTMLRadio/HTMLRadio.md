HTMLRadio example:

```js noeditor
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import HTMLRadio from './HTMLRadio';
```

```js
import Form from '../Form/Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import HTMLRadio from './HTMLRadio';


const onSubmit = async (values) => {
  console.log('HTMLRadio Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <HTMLRadio
      name="htmlRadioButton"
      label="RadioLabel1"
      value="RadioValue1"
      labelPlacement="bottom"
    />
    <HTMLRadio
      name="htmlRadioButton"
      label="RadioLabel2"
      value="RadioValue2"
      labelPlacement="top"
    />

    <MuiSubmit />
  </Form>
```
