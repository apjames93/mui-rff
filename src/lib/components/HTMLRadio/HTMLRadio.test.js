import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { HTMLRadio } from './HTMLRadio';
import Form from '../Form/Form';

function setup() {
  const props = {
    name: 'cool',
    value: 'htmlRadio1',
    label: 'apples',
  };
  const comp = mount(
    <Form onSubmit={() => true}>
      <HTMLRadio {...props} />
    </Form>,
  );
  return { comp, props };
}

describe('<HTMLRadio />', () => {
  it('renders HTMLRadio', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  it('simulate change and checkes that onchange toggles checked value', () => {
    const { props } = setup();

    const comp = mount(
      <Form onSubmit={() => true}>
        <HTMLRadio {...props} />
        <HTMLRadio {...props} value="woooo" />
      </Form>,
    );
    let radioInputProps = comp.find('[data-testid="HTMLRadio"]').first().props();
    expect(radioInputProps.checked).toBe(false);
    comp.find('[data-testid="HTMLRadio"]').first().simulate('change');
    radioInputProps = comp.find('[data-testid="HTMLRadio"]').first().props();
    expect(radioInputProps.checked).toBe(true);
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(
      <Form onSubmit={() => true}>
        <HTMLRadio {...props} />
      </Form>,
    );
    expect(tree).toMatchSnapshot();
  });
});
