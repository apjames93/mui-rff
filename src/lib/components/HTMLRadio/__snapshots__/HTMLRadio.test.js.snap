// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HTMLRadio /> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    className="MuiFormControlLabel-root"
    field={
      Object {
        "input": Object {
          "checked": false,
          "name": "cool",
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "type": "radio",
          "value": "htmlRadio1",
        },
        "meta": Object {
          "active": false,
          "data": Object {},
          "dirty": false,
          "dirtySinceLastSubmit": false,
          "error": undefined,
          "initial": undefined,
          "invalid": false,
          "length": undefined,
          "modified": false,
          "modifiedSinceLastSubmit": false,
          "pristine": true,
          "submitError": undefined,
          "submitFailed": false,
          "submitSucceeded": false,
          "submitting": false,
          "touched": false,
          "valid": true,
          "validating": false,
          "visited": false,
        },
      }
    }
    id="cool"
  >
    <input
      checked={false}
      data-testid="HTMLRadio"
      disabled={false}
      onChange={[Function]}
      type="radio"
      value="htmlRadio1"
    />
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      apples
    </span>
  </label>
</form>
`;
